File: ios-native/NeobankNative/Sources/Features/Auth/AuthController.swift
import Foundation
import Observation

/// Drives the auth lifecycle: reading the persisted session on launch, and
/// performing login/register/logout. Screens own their own submitting/error
/// state; this just owns the actions and the resulting session transition.
@MainActor
@Observable
final class AuthController {
    private let repository: AuthRepository
    private let tokenStorage: TokenStorage
    private let sessionStore: SessionStore

    init(repository: AuthRepository, tokenStorage: TokenStorage, sessionStore: SessionStore) {
        self.repository = repository
        self.tokenStorage = tokenStorage
        self.sessionStore = sessionStore
    }

    func bootstrap() {
        let hasToken = tokenStorage.readAccessToken() != nil
        sessionStore.setStatus(hasToken ? .authenticated : .unauthenticated)
    }

    func login(email: String, password: String) async throws {
        let tokens = try await repository.login(email: email, password: password)
        tokenStorage.saveTokens(accessToken: tokens.accessToken, refreshToken: tokens.refreshToken)
        sessionStore.setStatus(.authenticated)
    }

    func register(email: String, password: String, phone: String?, inviteCode: String?) async throws {
        let tokens = try await repository.register(
            email: email,
            password: password,
            phone: phone,
            inviteCode: inviteCode
        )
        tokenStorage.saveTokens(accessToken: tokens.accessToken, refreshToken: tokens.refreshToken)
        sessionStore.setStatus(.authenticated)
    }

    func logout() {
        tokenStorage.clear()
        sessionStore.setStatus(.unauthenticated)
    }
}

#if DEBUG
extension AuthController {
    static var preview: AuthController {
        let client = APIClient(baseURL: AppConfig.apiBaseURL)
        return AuthController(
            repository: AuthRepository(authClient: client, client: client),
            tokenStorage: TokenStorage(),
            sessionStore: SessionStore()
        )
    }
}
#endif

File: android-native/app/src/main/kotlin/dev/horobets/neobanknative/features/auth/AuthController.kt
package dev.horobets.neobanknative.features.auth

import androidx.compose.runtime.staticCompositionLocalOf
import dev.horobets.neobanknative.core.auth.SessionStatus
import dev.horobets.neobanknative.core.auth.SessionStore
import dev.horobets.neobanknative.core.storage.TokenStorage

/**
 * Drives the auth lifecycle: reading the persisted session on launch, and
 * performing login/register/logout. Screens own their own submitting/error
 * state; this just owns the actions and the resulting session transition.
 */
class AuthController(
    private val repository: AuthRepository,
    private val tokenStorage: TokenStorage,
    private val sessionStore: SessionStore,
) {
    fun bootstrap() {
        val hasToken = tokenStorage.readAccessToken() != null
        sessionStore.setStatus(if (hasToken) SessionStatus.AUTHENTICATED else SessionStatus.UNAUTHENTICATED)
    }

    suspend fun login(email: String, password: String) {
        val tokens = repository.login(email, password)
        tokenStorage.saveTokens(tokens.accessToken, tokens.refreshToken)
        sessionStore.setStatus(SessionStatus.AUTHENTICATED)
    }

    suspend fun register(email: String, password: String, phone: String?, inviteCode: String?) {
        val tokens = repository.register(email, password, phone, inviteCode)
        tokenStorage.saveTokens(tokens.accessToken, tokens.refreshToken)
        sessionStore.setStatus(SessionStatus.AUTHENTICATED)
    }

    fun logout() {
        tokenStorage.clear()
        sessionStore.setStatus(SessionStatus.UNAUTHENTICATED)
    }
}

/** Android's equivalent of SwiftUI's `.environment(_:)` / `@Environment(Type.self)`. */
val LocalAuthController = staticCompositionLocalOf<AuthController> {
    error("No AuthController provided — wrap the composition in AppEnvironment's providers")
}
